File: Capitulo-06/Pesquisa-Em-Largura/Exercicios/Typescript/breadth_first_search.js
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
function is_seller(name) {
    return name.endsWith("m");
}
function breadth_first_search(graph, start) {
    var queue = __spreadArray([], graph["you"], true);
    var verified = new Set();
    while (queue.length > 0) {
        var person = queue.shift();
        if (!verified.has(person)) {
            console.log("Verifiying ".concat(person));
            if (is_seller(person)) {
                console.log("".concat(person, " is seller"));
                return true;
            }
            else {
                queue.push.apply(queue, (graph[person] || []));
                verified.add(person);
            }
        }
    }
    return false;
}
var graph = {
    you: ["alice", "bob", "claire"],
    bob: ["anuj", "peggy"],
    alice: ["peggy"],
    claire: ["thom", "jonny"],
    anuj: [],
    peggy: [],
    thom: [],
    jonny: [],
};
breadth_first_search(graph, "you");

File: Capitulo-06/Pesquisa-Em-Largura/Exercicios/Ruby/breadth_first_search.rb
def is_seller(name)
  name.end_with? "m"
end

def breadth_first_search(graph, start)
  queue = graph[start].dup
  verified = []

  until queue.empty?
    person = queue.shift

    unless verified.include? person
      puts "Verifiying #{person}"

      if is_seller(person)
        puts "#{person} is seller"
        return true
      else
        queue.concat(graph[person] || [])
        verified << person
      end
    end
  end

  false
end

graph = {
  "you" => ["alice", "bob", "claire"],
  "bob" => ["anuj", "peggy"],
  "alice" => ["peggy"],
  "claire" => ["thom", "jonny"],
  "anuj" => [],
  "peggy" => [],
  "thom" => [],
  "jonny" => []
}

breadth_first_search(graph, "you")